### PR TITLE
Add ButinaSplitter class

### PR DESCRIPTION
**Pull request by proteneer**
This PR adds support for a ButinaSplitter that is similar to ScaffoldSplitting, except as opposed to using a hand-wavy Murcko Scaffold Decomposition it uses fingerprints. The goal here is similar to that of the ScaffoldSplitter: i.e. predict on new cores/chemotypes that are very dissimilar from that of the training set.

I'm going to run the benchmarks on this splitter and post results later.

### CODE REVIEW

**Comment by proteneer** [comment]
Note that this is a WIP: I'm not quite happy with how the scaffold/butina splitter class is doing the lumping. Currently it maximizes the train set with the biggest clusters. I think a better test (to better mimic real congeneric series in prospective studies) is to maximize the test/validation sets first.

**Comment by rbharath** [comment]
@proteneer Very cool! Glad to see this merged. I've certainly noted clumpiness in the Bemis-Murcko scaffolds (there will often be one giant cluster with 60% of the data and a long tail of little clusters). The PR looks good, and I'm glad to merge in once you think this is ready.

As a quick note the O(N^2) runtime might make Butina infeasible for the larger collections (PCBA has ~ 400K molecule). Might be worth us thinking about some approximation scheme which handles large datasets better.

**Comment by evanfeinberg** [comment]
So great to see this, @proteneer. Our field largely rests on the accuracy
of cross-validation for predicting performance on test, and this is a great
addition in our arsenal of non-random train-valid splits we can conduct.

**Comment by proteneer** [comment]
The Butina Clustering is giving some *really* interesting results:

```
(venv) yutong@symmetry:~/Code/deepchem/examples$ python benchmark.py -m graphconv -d tox21 -s butina
...
Performing butina clustering with cutoff of 0.18
# of actives per task: [  5.   6.   3.  29.  33.  21.   1.  43.  10.  11.  45.   7.]
...
computed_metrics: [0.8986331159119636, 0.95206280788177344, 0.90628111778513143, 0.90896348793685555, 0.78027133410372551, 0.88955350934539501, 0.93764019563723378, 0.82895267410436246, 0.93156379416611401, 0.8820863104606983, 0.90757527289371964, 0.9094853066647165]
computed_metrics: [0.59969969969969972, 0.99943310657596374, 0.77029914529914523, 0.73106111037145516, 0.73802338740098072, 0.8799838579499597, 0.1322033898305085, 0.53367750651355794, 0.97331189710610921, 0.71047430830039526, 0.82991092358180962, 0.90859140859140863]
```

```
(venv) yutong@symmetry:~/Code/deepchem/examples$ python benchmark.py -m graphconv -d tox21 -s butina
...
Performing butina clustering with cutoff of 0.22
# of actives per task in valid: [  3.   2.   5.  22.  25.  16.   4.  36.   6.  11.  29.   2.]
computed_metrics: [0.88694754311394564, 0.95264539740105958, 0.90693174280833744, 0.89526309308550278, 0.78127842374175138, 0.88935942288395298, 0.93282110807733065, 0.82587086456998338, 0.93294804648039276, 0.8780873770271449, 0.90400870277500311, 0.90725825898239698]
computed_metrics: [0.44958481613285883, 1.0, 0.82900763358778629, 0.79078455790784563, 0.67599999999999993, 0.89125251004016071, 0.72267206477732793, 0.67641559829059839, 0.97222222222222221, 0.53512396694214881, 0.86631433027204996, 0.70987654320987659]
```

**Comment by rbharath** [comment]
@proteneer Do you have a sense for what these numbers mean? (I see a `1.0` and a `.44` in there which is always strange)

**Comment by proteneer** [comment]
The validation set is pretty small. I've rewritten it so that the cluster joining is stopped as soon as there's at least 1 active for every test set. When there's very few actives (eg: [  3.   2.   5.  22.  25.  16.   4.  36.   6.  11.  29.   2.]) - it's very easy to get 1.0 AUCs.

This is done basically to mimic a true prospective-like series.

**Comment by proteneer** [comment]
@rbharath do you have any further thoughts on this?

**Comment by rbharath** [comment]
@proteneer Looks good to me. I'd be willing to merge once tests pass :-)

**Comment by proteneer** [comment]
rebased and squashed onto master, and updated docs. Waiting on tests now.

**Comment by rbharath** [comment]
@proteneer LGTM!
